sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/QueueReceiverAsyncClient.java

```diff
@@ -0,0 +1,303 @@
+        }
+        return openConsumer.get()
+            .receive()
+            .doOnCancel(() -> removeLink(linkName, SignalType.CANCEL))
+            .doOnComplete(() -> removeLink(linkName, SignalType.ON_COMPLETE))
+            .doOnError(error -> removeLink(linkName, SignalType.ON_ERROR));
+    }
+
+    private void removeLink(String linkName, SignalType signalType) {
+        logger.info("{}: Receiving completed. Signal[{}]", linkName, signalType);
+
+        if (openConsumer.get() != null) {
+            openConsumer.get().close();
+        }
+    }
+    private ServiceBusAsyncConsumer createServiceBusConsumer(String linkName, ReceiveMode receiveMode) {
+        final String entityPath = String.format(Locale.US, RECEIVER_ENTITY_PATH_FORMAT, getQueueName());
+
+        final Flux<AmqpReceiveLink> receiveLinkMono =
+            connectionProcessor.flatMap(connection ->
+                connection.createReceiveLink(linkName, entityPath, receiveMode))
+                .doOnNext(next -> logger.verbose("Creating consumer for path: {}", next.getEntityPath()))
+                .repeat();
+
+        final AmqpRetryPolicy retryPolicy = RetryUtil.getRetryPolicy(connectionProcessor.getRetryOptions());
+        final ServiceBusReceiveLinkProcessor linkMessageProcessor = receiveLinkMono.subscribeWith(
+            new ServiceBusReceiveLinkProcessor(prefetchCount, retryPolicy, connectionProcessor));
+
+        return new ServiceBusAsyncConsumer(linkMessageProcessor, messageSerializer, fullyQualifiedNamespace,
+            entityPath);
+    }
+
+    /**
+     * Abandon {@link Message} with lock token and updated message property. This will make the message available
+     * again for processing. Abandoning a message will increase the delivery count on the message.
+     *
+     * @param lockToken to be used.
+     * @param propertiesToModify Message properties to modify.
+     * @return The {@link Mono} the finishes this operation on service bus resource.
+     */
+    public Mono<Void> abandon(UUID lockToken, Map<String, Object> propertiesToModify) {
+        //TODO(feature-to-implement)
+        return null;
+    }
+
+    /**
+     * Abandon {@link Message} with lock token. This will make the message available again for processing.
+     * Abandoning a message will increase the delivery count on the message.
+     *
+     * @param lockToken to be used.
+     * @return The {@link Mono} the finishes this operation on service bus resource.
+     */
+    public Mono<Void> abandon(UUID lockToken) {
+        //TODO(feature-to-implement)
+        return null;
+    }
+
+    /**
+     * Completes a {@link Message} using its lock token. This will delete the message from the service.
+     *
+     * @param lockToken to be used.
+     * @return The {@link Mono} the finishes this operation on service bus resource.
+     */
+    public Mono<Void> complete(UUID lockToken) {
+        //TODO(feature-to-implement)
```

Comment:
In the JS stuff I just did I just took the entire message as the parameter. What do you think about that? I figure it requires the user to know less - just give me the whole object.

Reply:
I like the idea of "user know less". This is one place where all the languages can be same. Lets get it finalized during API review. One point to consider , does user have to cache whole object instead of id if they are dealing with multiple messages.